Bulk operations

graydon opened the pull request:
# Description

Transplant / reworking of @bobg's work in #1839 adapted to the new data layer. Also includes a change to inject entries into a LedgerTxn without loading, which improves catchup / bulk-load speed significantly.

# Checklist
- [x] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v5.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [x] Ran all tests
- [x] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

Perf: it goes much faster! Detailed measurements:


creates:
========
sqlite: 12726.1 entries/sec (single-entry writes, loading)
sqlite: 12998.4 entries/sec (single-entry writes, non-loading)
sqlite: 15543.3 entries/sec (bulk writes, loading)
sqlite: 20128.4 entries/sec (bulk writes, non-loading)

postgresql: 6905.01 entries/sec (single-entry writes, loading)
postgresql: 7304.49 entries/sec (single-entry writes, non-loading)
postgresql: 12127.3 entries/sec (bulk writes, loading)
postgresql: 33093.7 entries/sec (bulk writes, non-loading)

erases:
=======
sqlite: 13206.4 entries/sec (single-entry writes, loading)
sqlite: 15945.1 entries/sec (single-entry writes, non-loading)
sqlite: 17117.5 entries/sec (bulk writes, loading)
sqlite: 22613.9 entries/sec (bulk writes, non-loading)

postgresql: 8937.57 entries/sec (single-entry writes, loading)
postgresql: 19435.6 entries/sec (single-entry writes, non-loading)
postgresql: 13411.6 entries/sec (bulk writes, loading)
postgresql: 84095.5 entries/sec (bulk writes, non-loading)

MonsieurNicolas commented on src/ledger/LedgerTxnAccountSQL.cpp:
```diff
@@ -367,6 +370,348 @@ LedgerTxnRoot::Impl::deleteAccount(LedgerKey const& key)
+{
+
+    std::string sql =
+        "INSERT INTO accounts ( "
```
is there a difference between doing this and doing it the other way: `UPDATE ... ON CONFLICT ... INSERT` ?

graydon replied:
No idea. I'm not sure it actually works yet, still testing. Will try both.

graydon replied:
Turns out only the first exists; the second is not valid SQL. Because it's the world's most consistent language.

MonsieurNicolas commented on src/ledger/LedgerTxnAccountSQL.cpp:
```diff
@@ -367,6 +370,348 @@ LedgerTxnRoot::Impl::deleteAccount(LedgerKey const& key)
+marshalToPGArray(PGconn* conn, const std::vector<T>& v,
+                 const std::vector<soci::indicator>* ind = 0)
+{
+    std::string result = "{";
```
should you use a `ostringstream` here? this doesn't seem super efficient.

graydon replied:
Heh. I actually started with writing my own version with ostringstream and then decided it was needlessly fussing with Bob's code so copied his directly instead. But yeah I think that would be better too, I'll change it.

MonsieurNicolas commented on src/ledger/LedgerTxnAccountSQL.cpp:
```diff
@@ -367,6 +370,348 @@ LedgerTxnRoot::Impl::deleteAccount(LedgerKey const& key)
+postgresSpecificBulkUpsertAccounts(
+    Database& DB, std::vector<std::string> const& accountIDs,
+    std::vector<int64_t> const& balances, std::vector<int64_t> const& seqNums,
+    std::vector<uint32_t> const& subEntryNums,
```
I don't think we can use unsigned types in any SQL interfaces

graydon replied:
Ok. I'll fiddle with this (it's unsigned in the XDR but, ah right, I forgot we're doing this with the weird signed-with-non-negative-constraint thing in the SQL).

graydon replied:
fixed.

MonsieurNicolas commented on src/ledger/LedgerTxn.cpp:
```diff
@@ -1196,18 +1196,43 @@ LedgerTxnRoot::commitChild(EntryIterator iter)
+                // accounts with no extra signers at present.
+                if (iter.entry().data.account().signers.empty())
+                {
+                    AccountsToUpsert.push_back(iter.entry());
```
as keys are large, this may add non trivial overhead to ledgers that have a lot of ledger entries (related: the change https://github.com/stellar/stellar-core/pull/1871 turned a 5s ledger into sub second - all because of the change to keys); this is especially true for offers.


I wonder if we could instead use iterators for doing this as we are not doing anything interesting in accumulate other than very cheap filtering.

graydon replied:
Yeah, for the initial cut I wanted to be minimally invasive; a slightly-more-invasive version would hold copies of the iterators aside. I'm not sure it'll really be visible against the cost of growing giant string-y array literals to send as query parameters, but I'll do some measuring.

(Looking at #1871 I'm a little nervous about that -- did you ensure nothing map-order-sensitive was perturbed? Some things really like working over sets/maps in deterministic & sorted order.)

graydon replied:
fixed.

jonjove commented on src/ledger/LedgerTxnAccountSQL.cpp:
```diff
@@ -367,6 +370,348 @@ LedgerTxnRoot::Impl::deleteAccount(LedgerKey const& key)
+    std::string lastModifiedstr = marshalToPGArray(conn, lastModifieds);
+    std::string buyingLiabilitiestr =
+        marshalToPGArray(conn, buyingLiabilities, &liabilitiesInds);
+    std::string sellingLiabilitiestr =
```
Nitpick: could we do a naming scheme like `strSellingLiabilities` instead? Prevents weird mangling of names. For example, hard to remember if it is `sellingLiabilitiestr` or `sellingLiabilitiesstr` or `sellingLiabilitiesStr`.

jonjove commented on src/ledger/LedgerTxnAccountSQL.cpp:
```diff
@@ -367,6 +370,348 @@ LedgerTxnRoot::Impl::deleteAccount(LedgerKey const& key)
+    st.exchange(soci::use(accountIDs, "id"));
+    st.define_and_bind();
+    {
+        auto timer = DB.getUpsertTimer("account");
```
I think this should be `getDeleteTimer`.

graydon replied:
Yeah, copypasta, fixed.

graydon commented on src/ledger/LedgerTxn.cpp:
```diff
@@ -1196,18 +1196,43 @@ LedgerTxnRoot::commitChild(EntryIterator iter)
+                // present (less than 0.1% of accounts) and it is more
+                // complex to update signers in bulk, so we handle only
+                // accounts with no extra signers at present.
+                if (iter.entry().data.account().signers.empty())
```
Agreed! This was a source of some vexing bugs in my implementation today!

graydon commented on src/ledger/LedgerTxn.cpp:
```diff
@@ -1196,18 +1196,43 @@ LedgerTxnRoot::commitChild(EntryIterator iter)
 struct BulkLedgerEntryChangeAccumulator
```
Done.

graydon commented on src/ledger/LedgerTxn.cpp:
```diff
@@ -1196,18 +1196,43 @@ LedgerTxnRoot::commitChild(EntryIterator iter)
     void
     apply(LedgerTxnRoot::Impl& ltxr)
     {
+        ltxr.bulkUpsertAccounts(AccountsToUpsert);
```
Done.

graydon commented on src/ledger/LedgerTxn.cpp:
```diff
@@ -1196,18 +1196,43 @@ LedgerTxnRoot::commitChild(EntryIterator iter)
 struct BulkLedgerEntryChangeAccumulator
 {
 
+    std::vector<LedgerEntry> AccountsToUpsert;
```
Done.

graydon commented:
Updated with fixes and most-but-not-all review comments addressed. Plus bonus new accelerated non-loading variants that almost triple throughput again, above and beyond the batching. Still to do:

- [ ] Modify interface to not use unsigned types
- [ ] Modify accumulator to use iterators rather than entries
- [ ] Debug the experimental non-loading variants (and generally discuss design with others)
- [ ] Expand to cover other entry types (at least trustlines)

graydon commented:
Finished changes noted above, only remaining bits are data entries, deciding what to do about non-loading deletes, and maybe some extra comments?

vogel commented on src/ledger/LedgerTxnTrustLineSQL.cpp:
```diff
@@ -88,25 +100,11 @@ LedgerTxnRoot::Impl::insertOrUpdateTrustLine(LedgerEntry const& entry,
-                                 "outside of OperationFrame");
-    }
+    std::string accountIDStr, issuerStr, assetCodeStr;
+    getTrustLineStrings(tl.accountID, tl.asset, accountIDStr, issuerStr,
```
all of those looks like unrelated change, would you mind extracting to separate commit?

graydon replied:
fixed.

graydon commented:
@jonjove I believe this is now ready for review.

MonsieurNicolas commented on src/catchup/ApplyBucketsWork.cpp:
```diff
@@ -193,7 +197,8 @@ ApplyBucketsWork::advance(BucketApplicator& applicator)
             << formatSize(mAppliedSize) << "/" << formatSize(mTotalSize)
             << " in " << mAppliedBuckets << "/" << mTotalBuckets << " files ("
-            << (100 * mAppliedSize / mTotalSize) << "%)";
+            << (100 * mAppliedSize / mTotalSize) << "%, "
```
continuing conversation started in #1902 
> Continued in the bulk-operations PR #1906 . @MonsieurNicolas I felt it was helpful to get performance characteristics and since it's surfaced to field users as an INFO log message during catchup it might be an easily-compared signal about the difference in performance between our dev setups and the sort of IO subsystems our production users have. But if you feel strongly opposed I can drop it from that PR.

What I don't really like with this one is that it adds a new counter that cannot be reasoned about between runs (as it depends on which buckets got applied). We've been cleaning up metrics to allow people to import in bulk all metrics into systems like prometheus and this one fails the "usefulness" smell in a production environment.

If we want to give easy to read performance numbers in the logs, I would make the following changes:
* make the numbers easier to compare by reporting on a per bucket basis. Basically reset metrics when we start a new bucket, that way we can compare bucket 5.snap from one run with the same bucket 5.snap from a different run. 
* report only at the end of a bucket (INFO), optionally report every "advance" (in DEBUG)
* don't use the global medida registry (as to not export those metrics) or simple don't use medida metrics (ie: record the time)
* potentially break things down by bucket type (live vs delete) - I suspect deletes are adding a lot of noise

graydon replied:
done

graydon commented on src/ledger/LedgerTxn.cpp:
```diff
@@ -1180,22 +1305,121 @@ LedgerTxnRoot::Impl::throwIfChild() const
         for (; (bool)iter; ++iter)
         {
+            bulkApply(bleca);
+            if (bleca.accumulate(iter))
```
Mm. I'm a little hesitant to delete them all just now, though perhaps that's indicative of a lack of confidence in the new code. It's certainly been helpful during debugging to be able to `return false` on one of these accumulate branches and switch over to the entry-at-a-time code.

I guess you're right though that a finalization change that deletes the dead code and locks-in the new paths more visibly is an appropriate thing to add. If we find a problem, reverting that final change is no more perilous than re-enabling dead-code that's kept in the program text.

graydon replied:
fixed.

graydon commented on src/ledger/LedgerTxnAccountSQL.cpp:
```diff
@@ -294,6 +294,343 @@ LedgerTxnRoot::Impl::deleteAccount(LedgerKey const& key)
+        assert(e.entryExists());
+        assert(e.entry().data.type() == ACCOUNT);
+        auto const& account = e.entry().data.account();
+        accountIDs.push_back(KeyUtils::toStrKey(account.accountID));
```
done

graydon commented on src/ledger/LedgerTxn.h:
```diff
@@ -230,6 +230,8 @@ class EntryIterator
 
     LedgerKey const& key() const;
+
+    EntryIterator clone() const;
```
done

graydon commented on src/ledger/LedgerTxn.cpp:
```diff
@@ -1197,17 +1197,62 @@ LedgerTxnRoot::commitChild(EntryIterator iter)
+
+void
+LedgerTxnRoot::Impl::bulkApply(BulkLedgerEntryChangeAccumulator& bleca,
+                               size_t sizeLimit)
```
how about `bufferThreshold`?

graydon commented on src/ledger/LedgerTxn.h:
```diff
@@ -174,6 +174,27 @@ namespace stellar
+//
+//    - The count of rows in the database effected when applying the
+//      "erase" events might not be the expected number, so the consistency
+//      check we do there should be relaxed.
```
Mm. I initially thought this was not going to be a thing because an update or an insert effects the same number of rows at the db level. But it _does_ cause a similar problem in calculating LedgerChanges and LedgerTxnDeltas! Updated accordingly.

(Note: in fixing this, I did _not_ expand the consistency enum to a larger lattice of possible inconsistencies and their combinations; merely generalized/renamed the extra-deletes case to "inexact". While at first I thought this might mean we'd be weakening the rows-effected database-delete consistency check in some cases unnecessarily, thinking it over I think this is not an issue: such a weakened check would only be _unnecessary_ in cases where we were doing a bulk-upsert and _not_ a bulk-delete, but if that were the case then there'd _be_ no bulk-deletes to be weakening. IOW I think the only case we'd weaken _unnecessarily_ would, by definition, not be happening. Also it's somewhat of a corner case anyways, and there are a few ways it's already an overapproximation, not that it's ever very convincing to argue "I'm already being bad, why not be worse?", but TL;DR I don't think this is _actually_ any worse.)

graydon commented on src/ledger/LedgerTxnImpl.h:
```diff
@@ -267,13 +269,35 @@ class LedgerTxn::Impl
+    // eraseWithoutLoading has the basic exception safety guarantee. If it
+    // throws an exception, then
+    // - the entry cache in the parent LedgerTxnRoot may, but is not guaranteed
+    //   to, have the entry assiated with the given key removed.
```
fixed.

graydon commented on src/ledger/LedgerTxn.cpp:
```diff
@@ -1216,6 +1216,19 @@ BulkLedgerEntryChangeAccumulator::accumulate(EntryIterator const& iter)
     }
+    else if (iter.key().type() == TRUSTLINE)
+    {
+        if (iter.entryExists())
```
Fixed.

graydon commented on src/ledger/LedgerTxnTrustLineSQL.cpp:
```diff
@@ -182,6 +182,313 @@ LedgerTxnRoot::Impl::deleteTrustLine(LedgerKey const& key)
+        getTrustLineStrings(tl.accountID, tl.asset, accountIDStr, issuerStr,
+                            assetCodeStr);
+
+        accountIDs.push_back(accountIDStr);
```
fixed.

graydon commented on src/ledger/LedgerTxn.cpp:
```diff
@@ -265,6 +289,38 @@ LedgerTxn::Impl::create(LedgerTxn& self, LedgerEntry const& entry)
+    auto key = LedgerEntryKey(entry);
+
+    // dropFromEntryCacheIfPresent has strong exception safetly.
+    dropFromEntryCacheIfPresent(key);
```
I don't understand this comment. The function exists to support _non_-loading cases. Perhaps we can discuss tomorrow!

graydon commented on src/ledger/LedgerTxn.cpp:
```diff
@@ -524,6 +622,7 @@ LedgerTxn::getDelta()
 LedgerTxnDelta
 LedgerTxn::Impl::getDelta()
 {
+    throwIfNotExactConsistency();
```
Agreed. Fixed.

graydon commented on src/ledger/LedgerTxnAccountSQL.cpp:
```diff
@@ -297,359 +298,325 @@ LedgerTxnRoot::Impl::deleteAccount(LedgerKey const& key)
+LedgerTxnRoot::Impl::bulkUpsertAccounts(
+    std::vector<EntryIterator> const& entries)
+{
+  BulkUpsertAccountsOperation Op(mDatabase, entries);
```
fixed

graydon commented on src/ledger/LedgerTxnOfferSQL.cpp:
```diff
@@ -474,6 +503,357 @@ LedgerTxnRoot::Impl::deleteOffer(LedgerKey const& key)
+                      "unnest(:ids::BIGINT[]) "
+                      ") "
+                      "DELETE FROM offers WHERE "
+                      "(offerid) IN (SELECT * FROM r)";
```
just "consistency" / copypasta, no good reason. removed.

graydon commented on src/ledger/LedgerTxn.cpp:
```diff
@@ -1203,46 +1204,36 @@ LedgerTxnRoot::commitChild(EntryIterator iter)
+            mAccountDataToDelete.push_back(iter.clone());
+        break;
+    default:
+        assert(false);
```
fixed.

graydon commented on src/database/Database.cpp:
```diff
@@ -281,6 +285,22 @@ Database::isSqlite() const
+Database::doDatabaseTypeSpecificOperation(DatabaseTypeSpecificOperation &op)
+{
+  auto b = getSession().get_backend();
+  if (auto sq = dynamic_cast<soci::sqlite3_session_backend*>(b)) {
```
this is fixed later in the clang format run

graydon commented on src/ledger/LedgerTxnAccountSQL.cpp:
```diff
@@ -196,108 +196,6 @@ LedgerTxnRoot::Impl::writeSignersTableIntoAccountsTable()
 }
 
-void
-LedgerTxnRoot::Impl::insertOrUpdateAccount(LedgerEntry const& entry,
```
yeah, looking at this I think the existing code path is as good as we'd be getting if we tried to reuse the bulk-write pieces in the other order. current code issues 1 select for all loads and then N updates; reusing bulk-write would mean issuing N selects (and much bigger ones) followed by 1 bulk update. theoretically we can get a bit of a win if we have bulk-loading in the future but even then I don't know that it'd be worth bothering; it's a one-time action and updating a single column in a relatively sparse set of accounts (75k out of 2.3m)

graydon commented on src/ledger/LedgerTxnDataSQL.cpp:
```diff
@@ -55,95 +55,58 @@ LedgerTxnRoot::Impl::loadData(LedgerKey const& key) const
 
-void
-LedgerTxnRoot::Impl::insertOrUpdateData(LedgerEntry const& entry, bool isInsert)
+class BulkUpsertDataOperation : public DatabaseTypeSpecificOperation
```
done.

graydon commented on src/bucket/BucketApplicator.h:
```diff
@@ -25,9 +26,42 @@ class BucketApplicator
     BucketApplicator(Application& app, std::shared_ptr<const Bucket> bucket);
     operator bool() const;
-    size_t advance();
+    size_t advance(Counters *counters = nullptr);
```
sure; there's only one other (not very common) caller, so I just instantiated counters there too and passed by &. It's harmless.

graydon commented:
@jonjove spent the day cleaning this up / squashing / rebasing for readability; I cannot quite rebase onto master yet because it depends on some of the schema changes that got reverted, that I think we're planning on re-applying post-current-release. Otherwise I think ready for final review.

jonjove commented on src/ledger/LedgerTxnDataSQL.cpp:
```diff
@@ -410,17 +351,13 @@ LedgerTxnRoot::Impl::encodeDataNamesBase64()
-        {
-            CLOG(INFO, "Ledger") << "Wrote " << numUpdated << " data entries";
-        }
+        BulkUpsertDataOperation op(mDatabase, dataToEncode.begin(),
```
Looks like when I commented about this earlier, I was imagining it working differently from you. I was thinking of doing batches of size `LEDGER_ENTRY_BATCH_COMMIT_SIZE`, which would have been easier with iterators. I think I was making it a little more complicated than we needed, because the number of data entries is very small so I think we can do them in a single operation as you do here. No need to do it with iterators, might as well just pass `dataToEncode` by reference instead. Sorry for the confusion.

graydon replied:
Oh, ok! fixed to use plain vectors and a single-entry accumulate helper.

graydon commented on src/ledger/LedgerTxnImpl.h:
```diff
@@ -299,6 +318,7 @@ class LedgerTxnRoot::Impl
     mutable BestOffersCache mBestOffersCache;
     std::unique_ptr<soci::transaction> mTransaction;
     AbstractLedgerTxn* mChild;
+    LedgerTxnConsistency mConsistency;
```
The value is picked up and used in various leaf methods in the *SQL files. I could thread it through all those calls as a parameter, if you prefer?

graydon replied:
will use parameter

graydon commented on src/ledger/LedgerTxn.cpp:
```diff
@@ -1191,18 +1288,22 @@ LedgerTxnRoot::Impl::throwIfChild() const
     // guarantee, so use std::unique_ptr<...>::swap to achieve it
     auto childHeader = std::make_unique<LedgerHeader>(mChild->getHeader());
 
+    mConsistency = joinConsistencyLevels(mConsistency, cons);
```
Ah, true. Fixed.

graydon commented on src/ledger/LedgerTxn.cpp:
```diff
@@ -263,6 +287,35 @@ LedgerTxn::Impl::create(LedgerTxn& self, LedgerEntry const& entry)
+    throwIfSealed();
+    throwIfChild();
+
+    auto key = LedgerEntryKey(entry);
```
Hmm, dangling in the memory-unsafe sense? I thought the point of ltxe being a weak_ptr was that this would (correctly) trap when you try to access ltxe.current()?

graydon replied:
discussed in-chat, resolved.

graydon commented on src/ledger/LedgerTxn.cpp:
```diff
@@ -263,6 +287,35 @@ LedgerTxn::Impl::create(LedgerTxn& self, LedgerEntry const& entry)
+    auto key = LedgerEntryKey(entry);
+
+    auto current = std::make_shared<LedgerEntry>(entry);
+    auto impl = LedgerTxnEntry::makeSharedImpl(self, *current);
```
Do you mean this comment to apply to line 306, or the latter line 313 that contains `LedgerTxnEntry ltxe(impl)` ?

graydon commented on src/ledger/LedgerTxn.cpp:
```diff
@@ -263,6 +287,35 @@ LedgerTxn::Impl::create(LedgerTxn& self, LedgerEntry const& entry)
+    // can throw and the LedgerTxnEntry destructor requires that mActive
+    // contains key. LedgerTxnEntry constructor does not throw so this is
+    // still exception safe.
+    mActive.emplace(key, toEntryImplBase(impl));
```
Ah, it's only been a few weeks since I wrote the comment explaining what "active" means and I've already forgotten it / started to think it means something else. Ok.

graydon replied:
discussed in-chat, resolved.

graydon commented on src/ledger/LedgerTxn.cpp:
```diff
@@ -347,6 +400,46 @@ LedgerTxn::Impl::erase(LedgerKey const& key)
+    throwIfSealed();
+    throwIfChild();
+
+    mConsistency = LedgerTxnConsistency::EXTRA_DELETES;
```
nice catch. fixed.

graydon commented on src/ledger/LedgerTxn.cpp:
```diff
@@ -1191,18 +1288,22 @@ LedgerTxnRoot::Impl::throwIfChild() const
 
 void
 LedgerTxnRoot::Impl::commitChild(EntryIterator iter)
+void
```
oops! fixed.

graydon commented on src/ledger/LedgerTxnImpl.h:
```diff
@@ -267,13 +269,35 @@ class LedgerTxn::Impl
+    // If it throws an exception, then
+    // - the entry cache in the parent LedgerTxnRoot may, but is not guaranteed
+    //   to, have the given entry removed.
+    // - the current LedgerTxn::Impl may, but is not guaranteed to, have
```
I'm not sure what you mean here. Do you mean that I'm referring to the entry cache being cleared? that's residue from previous versions of this that did entry-cache clearing. I'll remove that. if you mean that it only presents the basic exception guarantee, I'm confused: create also only has the basic exception guarantee.

With the revised semantics of only populating the entry map (not changing the active map) suggests though that this can be made strong. Erase cannot, I don't think (assuming we want it to continue to affect both the active and entry maps).

graydon replied:
discussed in-chat, resolved.

graydon commented on src/ledger/LedgerTxnImpl.h:
```diff
@@ -267,13 +269,35 @@ class LedgerTxn::Impl
+    // throws an exception, then
+    // - the entry cache in the parent LedgerTxnRoot may, but is not guaranteed
+    //   to, have the entry assiated with the given key removed.
+    // - the current LedgerTxn::Impl may, but is not guaranteed to, have
```
discussed in-chat, resolved.

jonjove replied:
Think you may have missed actually resolving this one. Perhaps lost while rebasing?

graydon commented on src/ledger/LedgerTxnDataSQL.cpp:
```diff
@@ -112,6 +114,228 @@ LedgerTxnRoot::Impl::deleteData(LedgerKey const& key)
+                            std::vector<EntryIterator> const& entryIter)
+        : mDB(DB)
+    {
+        std::vector<LedgerEntry> entries;
```
yep, fixed.

graydon commented on src/ledger/LedgerTxnDataSQL.cpp:
```diff
@@ -112,6 +114,228 @@ LedgerTxnRoot::Impl::deleteData(LedgerKey const& key)
+
+  public:
+    void
+    accumulateEntries(std::vector<LedgerEntry>::const_iterator begin,
```
fixed.

graydon commented:
@jonjove updated (perhaps the last time?)

  - rebased atop revert-revert of the schema changes
  - added tests as requested
  - fixed all the exception safety, consistency-level and other smaller comments discussed today
  - added benchmarks and measured various states in the changeset (results in description above)

We're still hanging back from merging to master but I'll leave this PR here for you to rebase your stuff on if you like. If it's satisfactory I guess I'll also push it to graydon/db-perf-branch

jonjove commented on src/ledger/LedgerTxnDataSQL.cpp:
```diff
@@ -52,62 +55,222 @@ LedgerTxnRoot::Impl::loadData(LedgerKey const& key) const
+        : mDB(DB)
     {
-        throw std::runtime_error("could not update SQL");
+        std::vector<LedgerEntry> entries;
```
I think you are reserving space in this but not actually using it.

graydon replied:
Whoops! Dead code residue. Nice catch, fixed.

MonsieurNicolas commented on src/ledger/LedgerTxn.cpp:
```diff
@@ -1287,37 +1287,111 @@ LedgerTxnRoot::commitChild(EntryIterator iter, LedgerTxnConsistency cons)
+      std::vector<EntryIterator>& deleteBuffer)
+{
+    if (iter.entryExists())
+        upsertBuffer.emplace_back(iter);
```
maybe a question for later: should we accumulate in something like an `ordered_map` where the key follows the same rough order than the primary key in SQL? I think you mentioned that having things ordered was helping SQL quite a bit as it's touching the b-tree in a less random way

graydon replied:
Possible (or just sort before inserting) but on closer inspection & discussion with Jon I don't believe my earlier explanation of the performance of the synthetic benchmarks (having to do with order).

graydon commented on src/database/DatabaseTypeSpecificOperation.h:
```diff
@@ -0,0 +1,26 @@
+#pragma once
```
fixed.

graydon commented:
@MonsieurNicolas @jonjove added build fixes for MSVC and removal of dead code.

jonjove commented:
r+ 111b95f

jonjove commented:
r+ bb74f47
